engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,221 @@
+	StopControlSet
+
+	// StopControlCommenced indicates that stopping process has commenced and no parameters can be changed anymore.
+	// For example, blocks at or above stop height has been received, but finalization didn't reach stop height yet.
+	// It can only progress to StopControlPaused
+	StopControlCommenced
+
+	// StopControlPaused means EN has stopped processing blocks. It can happen by reaching set stop height, or
+	// if the node was started with a pause mode.
+	// It is a final state and cannot be changed
+	StopControlPaused
+)
+
+// NewStopControl creates new empty NewStopControl
+func NewStopControl(log zerolog.Logger, paused bool) *StopControl {
+	state := StopControlOff
+	if paused {
+		state = StopControlPaused
+	}
+	log.Debug().Msgf("created StopControl module with paused = %t", paused)
+	return &StopControl{
+		log:   log,
+		state: state,
+	}
+}
+
+// GetState returns current state of StopControl module
+func (s *StopControl) GetState() StopControlState {
+	return s.state
+}
+
+// IsPaused returns true is block execution has been paused
+func (s *StopControl) IsPaused() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.state == StopControlPaused
+}
+
+// SetStopHeight sets new stop height and crash mode, and return old values:
+//   - set, whether values were previously set
+//   - height
+//   - crash
+//
+// Returns error is the stopping process has already commenced, new values will be rejected.
+func (s *StopControl) SetStopHeight(height uint64, crash bool) (uint64, bool, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	oldHeight := s.height
+	oldCrash := s.crash
+
+	if s.state == StopControlCommenced {
+		return oldHeight, oldCrash, fmt.Errorf("cannot update stop height, stopping commenced for height %d with crash=%t", oldHeight, oldCrash)
+	}
+
+	if s.state == StopControlPaused {
+		return oldHeight, oldCrash, fmt.Errorf("cannot update stop height, already paused")
+	}
+
+	if s.lastExecutedHeight >= height {
+		return oldHeight, oldCrash, fmt.Errorf("cannot update stop height, given height %d at or below last executed %d", height, s.lastExecutedHeight)
+	}
+
+	s.log.Info().
+		Uint64("height", height).Bool("crash", crash).
+		Uint64("old_height", oldHeight).Bool("old_crash", oldCrash).Msg("new stop height set")
+
+	s.log.Debug().Int8("previous_state", int8(s.state)).Int8("new_state", int8(StopControlSet)).Msg("StopControl state transition")
+	s.state = StopControlSet
+
+	s.height = height
+	s.crash = crash
+	s.stopAfterExecuting = flow.ZeroID
+
+	return oldHeight, oldCrash, nil
+}
+
+// GetStopHeight returns:
+//   - height
+//   - crash
+//
+// Values are undefined if they were not previously set
+func (s *StopControl) GetStopHeight() (uint64, bool) {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.height, s.crash
+}
+
+// BlockProcessable should be called when new block is processable.
+// It returns boolean indicating if the block should be processed.
+func (s *StopControl) BlockProcessable(b *flow.Header) bool {
```

Comment:
At the moment, we are tracking the last executed block:
https://github.com/onflow/flow-go/blob/93680d1b7ad29944d05f85f3f36e85289cff2a0b/engine/execution/ingestion/stop_control.go#L25

But I am wondering if it wouldn't make sense to rather track the latest Block _scheduled_ for execution (`latestScheduledForExecution`). We would increase this variable here. 

The reasoning is that this is a completely asynchronous process. I think it makes sense to require for the admin command that only heights are specified that are not already scheduled for execution. Otherwise, we get into the complicated realm of "what happens if I set a height where the block execution is already in flight?"

Reply:
Let's keep discussion in one place https://github.com/onflow/flow-go/pull/3213#discussion_r993682312